File: src/components/DebugControls.cpp
#include "DebugControls.hpp"
#include "util/component_registry.hpp"
#include "core/Entity.hpp"
#include "core/Engine.hpp"
#include "graphics/RenderEngine.hpp"
#include "graphics/Camera.hpp"
#include "graphics/ImageEffect.hpp"
#include "input/Input.hpp"

REGISTER_COMPONENT_CLASS(DebugControls, "debugControls");

DebugControls::DebugControls(Entity* parent) : Component(parent), m_imgEffectsEnabled(true) { }

void DebugControls::start_impl()
{
	auto cam = Camera::main();
	if (cam) {
		m_imgEffects = cam->getEntity()->getComponents<ImageEffect>();
	}
}

void DebugControls::update_impl()
{
	Engine* engine = Engine::instance();
	RenderEngine* graphics = RenderEngine::instance();
	Input* input = Input::instance();

	if (input->getKeyPressed(key_escape)) {
		engine->stop();
	}

	if (input->getKeyPressed(key_p)) {
		bool d = !(graphics->isDeferredEnabled());
		graphics->setDeferredEnabled(d);
		std::cout << "Rendering path: " << (d ? "Deferred" : "Forward") << std::endl;
	}

	if (input->getKeyPressed(key_o)) {
		bool vfc = !(graphics->isVFCEnabled());
		graphics->setVFCEnabled(vfc);
		std::cout << "View frustum culling " << (vfc ? "Enabled" : "Disabled") << std::endl;
	}

	if (input->getKeyPressed(key_g)) {
		int om = int(graphics->getOutputMode());
		om = (om + 1) % RenderEngine::output_mode_max;
		graphics->setOutputMode(static_cast<RenderEngine::output_mode>(om));
		std::cout << "Debug output: " << getOutputName(om) << std::endl;
	}

	if (input->getKeyPressed(key_i)) {
		setImageEffectsEnabled(!m_imgEffectsEnabled);
		std::cout << "Image Effects " << (m_imgEffectsEnabled ? "Enabled" : "Disabled") << std::endl;
	}

	if (input->getKey(key_u)) {
		std::cout << "\r" << "FPS: " << engine->getFPS();
	}

	if (input->getKeyPressed(key_t)) {
		std::cout << "Triangle count: " << graphics->getTriangleCount() << std::endl;
	}

	if (input->getKey(key_r)) {
		engine->loadFirstScene();
	}
}

void DebugControls::setImageEffectsEnabled(bool val)
{
	if (val != m_imgEffectsEnabled) {
		m_imgEffectsEnabled = val;

		for (ImageEffect* e : m_imgEffects) {
			e->setEnabled(m_imgEffectsEnabled);
		}
	}
}

std::string DebugControls::getOutputName(int o)
{
	auto om = static_cast<RenderEngine::output_mode>(o);
	switch (om) {
	case RenderEngine::output_default:
		return "Default";
	case RenderEngine::output_diffuse:
		return "Diffuse Color";
	case RenderEngine::output_specular:
		return "Specular Color";
	case RenderEngine::output_smoothness:
		return "Smoothness";
	case RenderEngine::output_normal:
		return "Normals";
	case RenderEngine::output_depth:
		return "Depth";
	default:
		return "???";
	}
}

File: content/scripts/DebugControls.lua
DebugControls = Behaviour:new{
    imgEffectsEnabled = true
}

function DebugControls:init()
    local cam = Camera.main()
    if cam then
        self.imgEffects = cam:entity():getComponents("ImageEffect")
    end
end

function DebugControls:update()
    if Input.getKeyPressed("escape") then
        Engine.stop()
    end

    if Input.getKeyPressed("t") then
        print("Triangle count: "..Graphics.triangleCount())
    end

    if Input.getKeyPressed("p") then
        local d = not Graphics.isDeferredEnabled()
        Graphics.setDeferredEnabled(d)
        print("Rendering path: "..(d and "Deferred" or "Forward"))
    end

    if Input.getKeyPressed("o") then
        local vfc = not Graphics.isVFCEnabled()
        Graphics.setVFCEnabled(vfc)
        print("View frustum culling "..getEnabledStatus(vfc))
    end

    if Input.getKeyPressed("g") then
        local om = Graphics.outputMode()
        -- TODO: expose enums to lua
        om = (om + 1) % 6
        Graphics.setOutputMode(om)
        print("Debug output: "..getOutputName(om))
    end

    if Input.getKeyPressed("i") then
        self:setImgEffectsEnabled(not self.imgEffectsEnabled)
        print("Image Effects "..getEnabledStatus(self.imgEffectsEnabled))
    end

    if Input.getKey("u") then
        io.write("\rFPS: "..Engine.fps())
    end

    if Input.getKeyPressed("r") then
        Engine.loadFirstScene()
    end

    if Input.getKeyPressed("v") then
        local vs = not Engine.vSync()
        Engine.setVSync(vs)
        print("VSync "..getEnabledStatus(vs))
    end
end

function DebugControls:setImgEffectsEnabled(val)
    if (val ~= self.imgEffectsEnabled) then
        self.imgEffectsEnabled = val

        for i, e in ipairs(self.imgEffects) do
            e:setEnabled(self.imgEffectsEnabled)
        end
    end
end

omNames = { "Default", "Diffuse Color", "Specular Color", "Smoothness", "Normals", "Depth" }

function getOutputName(mode)
    return omNames[mode+1]
end

function getEnabledStatus(enabled)
    return enabled and "Enabled" or "Disabled"
end